### DISCUSSION TITLE
Unable to change network interface on VPU with 2 camera setup.

### DISCUSSION THREAD

**Discussion by matijazigic** [ifm/ifm3d#431] [Q&A] [answered]
Hello, 

I'm trying to change network interface on the VPU using command line or python script but without success. To be precise, I'm trying to change options for **_eth1_** to enable DHCP via command line and python script from the latest ifm docker stable image.
However, I'm able to change other parameters (device.info.name) but network stuff I cannot. 

Also, I'm able to change **_authorized_keys_**, and now I can connect to the VPU via eth0 and ssh key from my laptop.

Do you have any clue why?
Am I missing something?  
I did a reboot after changing, but nothing changed.

Here is the simple script:

```
import json
from ifm3dpy.device import O3R

o3r = O3R()

o3r.set({"device": {"info": {"name": "test4"}}})
o3r.set({"device": {"network": {"interfaces": {"eth1": {"useDHCP" : True} }}}})

print(o3r.get()["device"]["info"]["name"])
print(o3r.get()["device"]["network"]["interfaces"]["eth1"])

```

Output:

> test4
> {'ipv4': {'address': '192.168.42.69', 'dns': '0.0.0.0', 'gateway': '0.0.0.0', 'mask': 24}, 'mac': '00:02:01:43:7A:43', 'networkSpeed': 10, 'useDHCP': False}

Also, I have tried using ifmVisionAssistant, but result is the same. No error as output, but nothing changed.

![ifm_vision assist 1](https://github.com/ifm/ifm3d/assets/8402071/732574b4-9f04-404c-ba6b-2001218129fe)
![ifm vision assist 2](https://github.com/ifm/ifm3d/assets/8402071/f66fe0b7-ae75-457d-a158-af97d109d223)

**Comment by Stedag** [comment]
hi @matijazigic 

To set the specific interface to DHCP via iVA you would click the switch below the "DHCP" label, and then press the save button. This would automatically restart the VPU and then after reconnecting to the VPU in iVA or retrieving the config via the api, the changes should be visible.

I think the issue with the ifm3dpy set() call is that the "useDHCP" parameter expects a Boolean literal rather than an integer.

Using ifm3dpy 1.4.3 I'm getting when running the code above:
```
Error: Unknown error: /useDHCP: unexpected instance type; instance: 0
```

To set DHCP I used:
```
o3r.set({"device": {"network": {"interfaces": {"eth0": {"useDHCP": True,
}}}}})
o3r.reboot()
```

Hope this helps

**Comment by matijazigic** [comment]
Hello @Stedag,

thanks for quick answer.
I have copied wrong code, the right one is one you provided, but still with that code, I cannot change the interface. Also, I cannot do it via ifmVA, I tried numerous times, to click the switch, wait for apply and reset procedure, but after restarting iVA showing same nonDHCP options. So, in the end, it is all same unfortunately...

Do I need to factory restart the VPU maybe?

PS I'm trying to change eth1 not eth0.  And I'm also using ifm3d: version=1.4.3 from ifm docker. My firmware version is 1.1.30-1882.

**Comment by matijazigic** [comment]
Even after factory reset, I cannot switch to dhcp:

![ifm2](https://github.com/ifm/ifm3d/assets/8402071/f0234396-159a-49ed-b2e5-e055191ddc59)
![ifm1](https://github.com/ifm/ifm3d/assets/8402071/ee376d97-d176-420a-a71f-9e26311b891e)

While clicking on dhcp switch and save, there is a spinner, end with no error, and nothing changes afterwards, remains like on first two pictures.
![while_running_dhcp_and_reboot](https://github.com/ifm/ifm3d/assets/8402071/3210a6f5-042b-4eb0-9d1d-4e51525831af)

**Reply by graugans** [reply, marked as answer]
For the second Ethernet aka `eth1` there is no DHCP support enabled. To verify this one can query the Schema. Like this:

```sh
ovp8xx --ip 172.25.125.66 getSchema | jq .properties.device.properties.network.properties.interfaces.properties.eth1.properties.useDHCP
```

The result looks like this:
```JSON
{
  "default": false,
  "readOnly": true,
  "type": "boolean"
}
```

The parameter `useDHCP` is **readOnly** this is why the command `o3r.set({"device": {"network": {"interfaces": {"eth0": {"useDHCP": True,
}}}}})` does return successfully but no action has taken place. Any **readOnly** parameter which is set during a call to set is ignored by the device. This is a convenience function otherwise there would be additional filtering based on the Device schema required on the user side to remove all **readOnly** parameters.


Why is there no DHCP support for the second Ethernet? Because this may conflicts with the default routing of the first Ethernet in some scenarios. This is why we limited the second Ethernet to a static network configuration only. The intended use-case is to have this as a sensor network.

**Reply by lola-masson** [reply]
@matijazigic this seems to be a change we forgot to document in our release notes and documentation. I am updating the docs accordingly.

**Reply by matijazigic** [reply]
Hello @graugans, @lola-masson. Great support from you :)

I made an `eth0` as DHCP and leave `eth1` as it was. Now I can see changes on both sides (iVA and code). So now, I used ssh as `oem` user to VPU and logged successfully (via `eth1`). But I got problem while trying to ping outside network. I'm connected directly to the router with VPU `eth0`. When I try to connect that cable to, for example, my laptop, it normally connects. 
Do you know maybe what I'm doing wrong? My intention is to prepare VPU to pull some docker images in development phase using network and not sending by ssh these images. 

Setup:
![dhcp](https://github.com/ifm/ifm3d/assets/8402071/be4a0b53-51bd-4523-a18c-b142d2834bbf)
![dhcp2](https://github.com/ifm/ifm3d/assets/8402071/af160607-ca98-46b8-bca1-df7812cc27aa)

Unsuccessful ping:
![dhcp3](https://github.com/ifm/ifm3d/assets/8402071/c848a6f8-1781-485c-9d01-cfe355376fe3)

**Reply by graugans** [reply]
Maybe this is an issue with the DNS server not successful conveyed through the DHCP. Are you able to ping google.com from your host PC? In some environments there are Proxy Servers installed or firewalls installed which do block any outgoing traffic. 

To dig deeper into this I recommend to eliminate the DNS part and try to ping the Google DNS server first.

```sh
ping 8.8.8.8
```

If this works the traffic from your OVP80x is successfully routed to the internet. To verify the DNS settings one can use the command `systemd-resolve --status` on the OVP80x 

```sh
systemd-resolve --status | grep -A 10 "DNS Servers"
```

In case there is a routing issue the output of ` /sbin/route -n` is also of great help

**Reply by matijazigic** [reply]
Hello @graugans, once again thanks for the answer.

I tried what you said:

- from my host I can ping 8.8.8.8, 8.8.4.4, ... and google.com, so it is connected normally
- from VPU (address 192.168.100.157) I can ping my host (192.168.100.156) and my gateway (192.168.100.1), but _cannot_ 8.8.8.8, .... , google.com
- I have generated reports from _systemd-resolve_ and _route_

[settings.txt](https://github.com/ifm/ifm3d/files/13590789/settings.txt)
[settings_dns.txt](https://github.com/ifm/ifm3d/files/13590792/settings_dns.txt)
[route.txt](https://github.com/ifm/ifm3d/files/13590793/route.txt)

iVA changed automatically address of VPU to 192.168.100.157 once started. 

![temp_address](https://github.com/ifm/ifm3d/assets/8402071/bf51b286-a267-4796-9cc7-83fffa87dd07)

Ping google 
![ping](https://github.com/ifm/ifm3d/assets/8402071/0c9edd54-9745-4c6e-934a-751487d907b9)

Ping host and gateway 
![ping_gateway_and_dns](https://github.com/ifm/ifm3d/assets/8402071/65baecd5-18e6-45a1-b96c-2390177e4d3c)

My ethernet network on host
![my_address](https://github.com/ifm/ifm3d/assets/8402071/e207f43c-feab-48e7-bea6-4e1b5c9019a0)

----------------------

Afterwards I tried to disconnect and connect again and now there is much more inside commands output. 

[settings.txt](https://github.com/ifm/ifm3d/files/13591027/settings.txt)
[settings_dns.txt](https://github.com/ifm/ifm3d/files/13591029/settings_dns.txt)
[route.txt](https://github.com/ifm/ifm3d/files/13591026/route.txt)

But once again there is connection to the host and gateway/dns, but no ping to the 8.8.8.8, and google. **Now ping to those addresses doesn't return error just hangs**. So still no luck with the connection to the outside world. 

![new_ping](https://github.com/ifm/ifm3d/assets/8402071/5e486034-27e0-4a17-9666-a1a4180a23b6)


iVA now looks like

![image](https://github.com/ifm/ifm3d/assets/8402071/7e7724ba-54c4-451c-a609-e3e0b77a3af9)

One more detail, I'm using my home network in this setup, so VPU is connected directly to the router. Only my host/laptop, from which I'm using iVA is behind company VPN, and connected to the router network.

**Reply by graugans** [reply]
I have to check this at home with my router. Within the company we are blocked behind a cooperate firewall and a proxy server.

**Reply by graugans** [reply]
I can confirm this behavior there is an issue in combination with the link-local address fallback and the DHCP configuration. When your ISP supports IPv6 at least an IPv6 ping is possible.

```sh
root@ovp8xx-54-f7-37:~# ping www.google.com
PING www.google.com (142.250.186.68): 56 data bytes
^C
--- www.google.com ping statistics ---
4 packets transmitted, 0 packets received, 100% packet loss
root@ovp8xx-54-f7-37:~# ping6 www.google.com
PING www.google.com (2a00:1450:4001:828::2004): 56 data bytes
64 bytes from 2a00:1450:4001:828::2004: seq=0 ttl=115 time=40.308 ms
64 bytes from 2a00:1450:4001:828::2004: seq=1 ttl=115 time=37.422 ms
```

**Reply by matijazigic** [reply]
So, just to confirm, with such ISP, DHCP option is not fully operational?

**Reply by graugans** [reply]
It is not an ISP issue it is a configuration issue on our device, I am currently checking if there is a possible workaround.

**Reply by graugans** [reply]
Sorry for the delay, I have checked a possible workaround for your issue. I can access the internet using a static IP from my home router. 

```json
{"device":{"network":{"interfaces":{"eth0":{"ipv4":{"address":"192.168.178.42","dns":"192.168.178.1","gateway":"192.168.178.1","mask":24},"mac":"48:B0:2D:54:F7:37","networkSpeed":1000,"useDHCP":false}}}}}
```
With this in place, I can pull the image

```sh
ovp8xx-54-f7-37:~$ docker pull ubuntu:20.04
20.04: Pulling from library/ubuntu
dae58cbd668a: Pull complete 
Digest: sha256:f5c3e53367f142fab0b49908550bdcdc4fb619d2f61ec1dfa60d26e0d59ac9e7
Status: Downloaded newer image for ubuntu:20.04
docker.io/library/ubuntu:20.04
```

**Reply by graugans** [reply]
With such an embedded device certificate handling might be outdated when not updating the firmware. But it looks like you are safe for some time:

```sh
ovp8xx-54-f7-37:~$ DOM="docker.io"
ovp8xx-54-f7-37:~$ PORT="443"
ovp8xx-54-f7-37:~$ printf Q | openssl s_client -servername $DOM -connect $DOM:$PORT | openssl x509 -noout -dates
depth=2 C = US, O = Amazon, CN = Amazon Root CA 1
verify return:1
depth=1 C = US, O = Amazon, CN = Amazon RSA 2048 M01
verify return:1
depth=0 CN = *.docker.com
verify return:1
DONE
notBefore=Oct  2 00:00:00 2023 GMT
notAfter=Oct 31 23:59:59 2024 GMT
```
